bftengine/src/bftengine/KeyExchangeManager.cpp

```diff
@@ -76,181 +58,95 @@ std::string KeyExchangeManager::generateCid() {
-  }
-  LOG_INFO(KEY_EX_LOG, "Check point  " << num << " trigerred rotation ");
-  LOG_INFO(KEY_EX_LOG, "building crypto system ");
-  notifyRegistry(true);
-}
-
-void KeyExchangeManager::registerForNotification(IKeyExchanger* ke) { registryToExchange_.push_back(ke); }
-
-KeyExchangeMsg KeyExchangeManager::getReplicaPublicKey(const uint16_t& repID) const {
-  return keyStore_.getReplicaPublicKey(repID);
+void KeyExchangeManager::loadPublicKeys() {
+  // after State Transfer public keys for all replicas are expected to exist
+  auto num_loaded = publicKeys_.loadAllReplicasKeyStoresFromReservedPages();
+  if (ReplicaConfig::instance().getkeyExchangeOnStart()) ConcordAssert(num_loaded == clusterSize_);
+  LOG_INFO(KEY_EX_LOG, "building crypto system after state transfer");
+  notifyRegistry();
 }
 
-// Called at the end of a state transfer.
-void KeyExchangeManager::loadKeysFromReservedPages() {
-  // Until Key-rotation is implemented, state transfer after the key-exchange has been performed has no side effect.
-  // in order to mitigate BC-5530, we'll not do a redundent work
-  if (keysExchanged) {  // TODO remove when implementing key-rotation!
-    LOG_INFO(KEY_EX_LOG, "After state transfer, return due to already updated crypto system");
+void KeyExchangeManager::sendKeyExchange(const SeqNum& sn) {
+  // first check whether we've already generated keys lately
+  if (private_keys_.lastGeneratedSeqnum() &&  // if not init  ial
+      (sn - private_keys_.lastGeneratedSeqnum()) / checkpointWindowSize < 2) {
+    LOG_INFO(KEY_EX_LOG, "ignore request - already generated keys for seqnum: " << private_keys_.lastGeneratedSeqnum());
     return;
   }
-  // If we reached state transfer, all keys should have beed exchanged in the source replica
-  // TODO might be changed when full rotation is imp
-  ConcordAssert(keyStore_.loadAllReplicasKeyStoresFromReservedPages());
-  // TODO E.L  KeyRotation implementation will need to rotate privete keys here, checkpoint number will be essential
-  if (keysView_.keys().privateKey.empty()) {
-    if (keysView_.keys().outstandingPrivateKey.size() > 0) {
-      keysView_.rotate(keysView_.keys().privateKey, keysView_.keys().outstandingPrivateKey);
-    } else {
-      keysView_.rotate(keysView_.keys().privateKey, keysView_.keys().generatedPrivateKey);
-    }
-    ConcordAssert(keysView_.keys().privateKey.empty() == false);
-  }
-  keysExchanged = true;
-  LOG_INFO(KEY_EX_LOG, "building crypto system after state transfer");
-  notifyRegistry(true);
-}
-
-void KeyExchangeManager::sendKeyExchange() {
   KeyExchangeMsg msg;
+  auto cid = generateCid();
   auto [prv, pub] = multiSigKeyHdlr_->generateMultisigKeyPair();
-  keysView_.keys().generatedPrivateKey = prv;
-  msg.key = pub;
+  private_keys_.key_data().generated.priv = prv;
+  private_keys_.key_data().generated.pub = pub;
+  private_keys_.key_data().generated.cid = cid;
+  private_keys_.key_data().generated.sn = sn;
+  private_keys_.save();
```

Comment:
Do we have an issue if for some reason we fail to send but save the file, in the prev imp I saved the file after sending, not sure

Reply:
In this case the system won't be able to generate and exchange keys for two checkpoints, but will continue to operate normally with the old keys.
In your case there could be a problem in case of failure to persist generated keys after sending them to consensus.